File: Data Structures & Algorithms/stone-game/submission-2.cpp
class Solution {
public:
    int solve(vector<int>& piles,int left,int right,auto& dp){
        if(left > right)return 0;
        if(dp[left][right] != -1)return dp[left][right];
        int takeLeft = piles[left] - solve(piles,left+1,right,dp);
        int takeRight = piles[right] - solve(piles,left,right-1,dp);
        return dp[left][right] = max(takeLeft,takeRight);
    }
    bool stoneGame(vector<int>& piles) {
        vector<vector<int>> dp(piles.size(),vector<int>(piles.size(),-1));
        return solve(piles,0,piles.size()-1,dp) > 0;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int solve(TreeNode root,int maxi){
        if(root == null)return 0;
        maxi = Math.max(root.val,maxi);
        int left = solve(root.left,maxi);
        int right = solve(root.right,maxi);
        return left + right + ((root.val >= maxi) ? 1 : 0);
    }
    public int goodNodes(TreeNode root) {
        return solve(root,root.val);
    }
}
